### PR TITLE
[windows/arm64] Add missing JDK toolchain for java build

### PR DESCRIPTION
**Pull request by nsait-linaro**
Extend configurations to add JDK 11 and 17 for windows/arm64 platforms.

This should fix the Java builds on windows/arm64

### CODE REVIEW

**Comment by nsait-linaro** [comment]
@meteorcloudy and @philwo - Can you please review this change? This would be needed for the java builds.

File: WORKSPACE [hunk 1]
```diff
@@ -637,6 +653,20 @@ exports_files(["WORKSPACE"], visibility = ["//visibility:public"])
...
+http_archive(
+    name = "openjdk11_windows_arm64_archive",
+    build_file_content = """
+java_runtime(name = 'runtime', srcs =  glob(['**']), visibility = ['//visibility:public'])
+exports_files(["WORKSPACE"], visibility = ["//visibility:public"])
+""",
+    sha256 = "b8a28e6e767d90acf793ea6f5bed0bb595ba0ba5ebdf8b99f395266161e53ec2",
+    strip_prefix = "jdk-11.0.13+8",
+    urls = [
+        "https://aka.ms/download-jdk/microsoft-jdk-11.0.13.8.1-windows-aarch64.zip"
```
Comment (philwo) [inline WORKSPACE:666]: Can you add this mirror URL before the upstream one, please? I just uploaded it there:

```https://mirror.bazel.build/aka.ms/download-jdk/microsoft-jdk-11.0.13.8.1-windows-aarch64.zip```

File: distdir_deps.bzl [hunk 2]
```diff
@@ -440,6 +440,19 @@ DIST_DEPS = {
...
+    "remotejdk11_win_arm64": {
+        "aliases": [
+            "remotejdk11_win_arm64_for_testing",
+            "openjdk11_windows_arm64_archive",
+        ],
+        "archive": "microsoft-jdk-11.0.13.8.1-windows-aarch64.zip",
+        "sha256": "b8a28e6e767d90acf793ea6f5bed0bb595ba0ba5ebdf8b99f395266161e53ec2",
+        "strip_prefix": "jdk-11.0.13+8",
+        "urls": [
+            "https://aka.ms/download-jdk/microsoft-jdk-11.0.13.8.1-windows-aarch64.zip",
```
Comment (philwo) [inline distdir_deps.bzl:452]: same mirror URL here

**Review by philwo — APPROVED** [review]

**Review by philwo — APPROVED** [review]

**Comment by philwo** [comment]
Huh, not sure why the mirror URL is giving HTTP 404... let me check.

**Comment by nsait-linaro** [comment]
> Huh, not sure why the mirror URL is giving HTTP 404... let me check.
> 
> Edit: Fixed! I'll rerun the failed CI jobs.

Thanks! Looks like the failing job also has trouble downloading JDK package for darwin from Bazel mirror. If this is already fixed or just an intermittent issue, could you retrigger the build please ?

```
test_verify_urls FAILED: Invalid urls: https://mirror.bazel.build/zulu/bin/zulu17.30.15-ca-jdk17.0.1-macosx_aarch64.tar.gz.
/b/f/w/bazel-out/k8-fastbuild/bin/src/test/shell/bazel/verify_workspace.runfiles/io_bazel/src/test/shell/bazel/verify_workspace:62: in call to test_verify_urls
INFO[verify_workspace 2022-02-03 15:11:37 (+0000)] Cleaning up workspace

```

**Comment by philwo** [comment]
@nsait-linaro I'm really sorry, I don't know why this test is suddenly failing with this error. The URL seems fine and I can download it on my own machine. I'll retry it later today, maybe it's some really weird kind of network flakyness between RBE and the remote server? I don't see anything in your PR that would cause it. @meteorcloudy Any idea?

**Comment by meteorcloudy** [comment]
Hmm, it does seem to be flaky in another presubmit: https://buildkite.com/bazel/bazel-bazel-github-presubmit/builds/10200#f7c14b0e-60b7-4f4f-9440-e78c19da31ef

But not sure what caused it..

**Comment by nsait-linaro** [comment]
> @nsait-linaro I'm really sorry, I don't know why this test is suddenly failing with this error. The URL seems fine and I can download it on my own machine. I'll retry it later today, maybe it's some really weird kind of network flakyness between RBE and the remote server? I don't see anything in your PR that would cause it. @meteorcloudy Any idea?

Thanks, @philwo. Could you do one more retrigger please ?

**Comment by comius** [comment]
> Hmm, it does seem to be flaky in another presubmit: https://buildkite.com/bazel/bazel-bazel-github-presubmit/builds/10200#f7c14b0e-60b7-4f4f-9440-e78c19da31ef
> 
> But not sure what caused it..

I noticed in this in one of my PRs before. At that PR it looked like we're hitting some kind of traffic limiter - one url in the list consistently failed remotely (i'm only assuming it was the last one), and the same url didn't fail locally.

**Comment by philwo** [comment]
I'll submit a change to make this test easier to debug, as we currently have no idea or log output why it's failing. :/

@comius That's a good idea. If it's indeed a traffic limit, maybe RBE team can help us out, as this is only failing on RBE, so it must be something specific to the environment.

**Comment by philwo** [comment]
@nsait-linaro Could you please rebase this PR against master to pick-up https://github.com/bazelbuild/bazel/commit/938e20947d7f51819ab292e1569c62aaecd73d33? I'm curious if this will tell us why the test is failing.

**Comment by nsait-linaro** [comment]
> @nsait-linaro Could you please rebase this PR against master to pick-up [938e209](https://github.com/bazelbuild/bazel/commit/938e20947d7f51819ab292e1569c62aaecd73d33)? I'm curious if this will tell us why the test is failing.



> @nsait-linaro Could you please rebase this PR against master to pick-up [938e209](https://github.com/bazelbuild/bazel/commit/938e20947d7f51819ab292e1569c62aaecd73d33)? I'm curious if this will tell us why the test is failing.

Looks like still timing out

```
Checking https://mirror.bazel.build/zulu/bin/zulu17.30.15-ca-jdk17.0.1-macosx_aarch64.tar.gz ...
curl: (7) Failed to connect to mirror.bazel.build port 443: Connection timed out

```

**Comment by philwo** [comment]
Nice, at least now we get an error message. I'll ask the RBE folks for help.

**Comment by philwo** [comment]
I filed a bug asking for help, will follow-up as soon as I hear back!

**Comment by nsait-linaro** [comment]
> I filed a bug asking for help, will follow-up as soon as I hear back!

@philwo Is CI issue specific to this patch ? I just retriggered and still seems to have the same issue

**Comment by meteorcloudy** [comment]
@nsait-linaro Yes, because this PR adds more URLs, which I think happen to triggered some traffic limit.

**Comment by meteorcloudy** [comment]
While the RBE team is looking into this (which looks like going to take some time), one mitigation I can think of is to deduplicate the urls we parsed from files:
https://cs.opensource.google/bazel/bazel/+/master:src/test/shell/bazel/verify_workspace.sh;l=50, so that we have less curl requests.

**Comment by philwo** [comment]
@meteorcloudy Yes, I also looked into this and one easy win would be to update all our other OpenJDKs to the latest version - we have at least one case where we have two separate versions for the same platform in the file. This should save some URLs 😅

**Comment by meteorcloudy** [comment]
> Deduping the list would be a good improvement for the test anyway, great idea!

@nsait-linaro Do you mind trying to fix the [verify_workspace](https://cs.opensource.google/bazel/bazel/+/master:src/test/shell/bazel/verify_workspace.sh;l=50) test? I'm quite overloaded currently.

**Comment by nsait-linaro** [comment]
> > Deduping the list would be a good improvement for the test anyway, great idea!
> 
> @nsait-linaro Do you mind trying to fix the [verify_workspace](https://cs.opensource.google/bazel/bazel/+/master:src/test/shell/bazel/verify_workspace.sh;l=50) test? I'm quite overloaded currently.

Yes of course!

**Comment by nsait-linaro** [comment]
> > > Deduping the list would be a good improvement for the test anyway, great idea!
> > 
> > 
> > @nsait-linaro Do you mind trying to fix the [verify_workspace](https://cs.opensource.google/bazel/bazel/+/master:src/test/shell/bazel/verify_workspace.sh;l=50) test? I'm quite overloaded currently.
> 
> Yes of course!

Hopefully, https://github.com/bazelbuild/bazel/pull/14763/files should fix!

File: WORKSPACE [hunk 3]
```diff
@@ -231,9 +231,11 @@ dist_http_file(
     downloaded_file_path = "zulu-win-vanilla.zip",
 )
 
-dist_http_file(
+http_file(
```
Comment (meteorcloudy) [inline WORKSPACE:234]: Can we keep using dist_http_file and move the definition into distdir_deps.bzl?

File: distdir_deps.bzl [hunk 4]
```diff
@@ -460,6 +461,19 @@ DIST_DEPS = {
...
+    "remotejdk11_win_arm64": {
+        "aliases": [
+            "remotejdk11_win_arm64_for_testing",
+            "openjdk11_windows_arm64_archive",
+        ],
+        "archive": "microsoft-jdk-11.0.13.8.1-windows-aarch64.zip",
+        "sha256": "b8a28e6e767d90acf793ea6f5bed0bb595ba0ba5ebdf8b99f395266161e53ec2",
+        "strip_prefix": "jdk-11.0.13+8",
+        "urls": [
+            "https://mirror.bazel.build/aka.ms/download-jdk/microsoft-jdk-11.0.13.8.1-windows-aarch64.zip",
```
[the remark this answers is not in the document]
Comment (nsait-linaro) [reply distdir_deps.bzl:473]: Yes, Azul doesn't seem to have binaries for arm64 JDK11. So we had to use the MS builds.

File: WORKSPACE (hunk 3 above)
Comment (nsait-linaro) [reply WORKSPACE]: Done !

**Review by meteorcloudy — APPROVED** [review]

**Comment by nsait-linaro** [comment]
Thanks, @meteorcloudy, and @philwo.

Looking for the per-commit build for validation, Is it still at the same location? I've tried with the latest commit but cannot find anything there.

https://storage.googleapis.com/bazel-builds/artifacts/windows_arm64/b6024258c5b9b396b022bad6e7e3d64423048289/bazel.exe

**Comment by nsait-linaro** [comment]
Thanks, @meteorcloudy, I can confirm that java and c++ builds are working fine now. It will be nice to get this delivered with the next Bazel release with the `.exe` suffix :-)

**Comment by meteorcloudy** [comment]
@bazel-io fork 5.1

**Comment by meteorcloudy** [comment]
@nsait-linaro That's nice! Can you send a PR to cherry-pick Windows ARM64 related changes to Bazel 5.1? I have opened issues to track the progress.

> the next Bazel release with the .exe suffix

Yes, the official release will of course come with ".exe" suffix.

**Pull request closed without merging**